README.md
# Friendly_Gadget
ALX-Project_Portfolio
# Friendly Gadget
![GitHub repo size](https://img.shields.io/github/repo-size/Yusuf-R/Friendly_Gadget)
![GitHub issues](https://img.shields.io/github/issues/Yusuf-R/Friendly_Gadget)
![GitHub Repo stars](https://img.shields.io/github/stars/Yusuf-R/Friendly_Gadget?logo=github&style=flat)
![GitHub forks](https://img.shields.io/github/forks/Yusuf-R/Friendly_Gadget?logo=github&style=falt)
![GitHub commit activity](https://img.shields.io/github/commit-activity/m/Yusuf-R/Friendly_Gadget?logo=github)

![banner](web_dynamic/static/assets/img/cover.png)

## Overview

Friendly Gadget is a web-based platform designed to help people understand your gadget choices. It will be your one point of reference for a complete informative guide about a product or service of your mobile or computer devices

## The Story
Introducing our revolutionary web application FriendlyGadget, designed to transform the way you choose your mobile devices. Picture this: a world where selecting the perfect device is a seamless, informed, and exhilarating experience. It all began with a realization - a common thread among conversations with friends, colleagues, and peers. The need for a mobile device, be it an upgrade or a fresh start, was apparent. Yet, there was always a gap, a disconnect between understanding the features of a device and envisioning its optimal performance in real-world applications.

I vividly recall the quest to find the ideal mobile device for my dear friend. This decision was monumental, meant to cater to her every desire, encompassing aesthetics, performance, and application capabilities. The journey through countless device options was akin to a thrilling rollercoaster ride. With so many choices and limited knowledge of how their features would fare in diverse application scenarios, the process became arduous. It demanded time, consideration, and meticulous evaluation of various factors.

Finally, after extensive deliberation, we arrived at a decision. Yet, as the dust settled, I couldn't help but reflect on the time-consuming nature of the process. The pace of decision-making was nerve-wracking thus, we need a solution to help Fastrack user decision.

This led us to create a groundbreaking solution FreindlyGadget - a web-based application that streamlines access to vital information and application resources. Say goodbye to hours of research and uncertainty. Say hello to a future where confidence meets efficiency. Welcome to our game-changing mobile device selection experience.

Front End
* HTML
* CSS
* JavaScript
* BoostStrap CSS Framework for developing responsive and mobile-first websites.

RESTful API
* GET, POST, PUT and DELETE requests handled
* CRUD manipulation through RESTful API

Relational Database
* MySql Handled with ORM (SQLAlchemy)
* Model system with base model handling identification
* Many to many relationships for users and rewards

Server / Deployment
* Nginx / Gunicorn
* Ubuntu 20.04

## Features

The Friendly_Gadget APP includes the following key features:

1. **Get the Best Device:** Users can get the device that is suitable for their purposes with ease, by displaying the search results based on their requirements

2. **Identify your Device:** Users can view the device's specifications and capabilities and identify each characteristic separately.




# Installation
## Dependencies
This application is written in Python and requires Python 3.8 or later to run correctly.

## Getting Started

Clone this git repository. If you are a GNU/Linux user, you could copy and paste the
following command to clone and change the working directory into the root of this project:

```sh
git clone https://github.com/Yusuf-R/Friendly_Gadget && cd Friendly_Gadget
```

Otherwise, clone the repository as you'd like and change the working directory into
the root of the project.
## Command Interpreter (The Console)

The command interpreter is a command-line interface that allows users to interact
with the **Freindly_Gadget** system. It provides a set of commands for managing the various
aspects of the system, such as creating and managing Brands, Models and so on.

### Starting the Command Interpreter

To start the command interpreter, navigate to the root of the project if not already there
and run the following command:

```sh
./console.py
```

This will start the command interpreter and you will be presented with a prompt,
`[FreindlyGadget]: `, where you can start entering commands.

### Using the Command Interpreter

The command interpreter provides a set of commands for managing the various aspects
of the **Friendly_Gadget** clone system. Some of the available commands include:

| Command         | Description                                                       | Usage                                                                 |
| ---------       | ----------------------------------------------------------------- | --------------------------------------------------------------------- |
| `create_model`  | Create a new instance of a class                                  | `create_model <class_name>`                                           |
| `show`          | Show the details of a specific instance of a class                | `show <class_name> <instance_id>`                                     |
| `all`           | Display the details of all instances or all instances of a class  | `all` or `all <class_name>`                                           |
| `delete`        | Delete a specific instance of a class                             | `delete <class_name> <instance_id>`                                   |
| `update`        | Update the attribute value for a given instance of a class        | `update <class_name> <class_id> <attribute_name> "<attribute_value>"` |
| `quit`          | Quits the command interpreter                                     | `quit`                                                                |                                                                   |

> The interpreter can also be terminated by hitting the `EOF` key combination.

To use a command, simply type the command followed by any required arguments at the prompt
and hit enter.

For example, to create a new instance of the `Brand` class, you would run the following command:

```sh
[FreindlyGadget]: create_brand Apple
```

To show the details of a specific instance of a class, you would run the following command,
where `<class_id>` is the ID of the instance you want to show:

```sh
[FreindlyGadget]: show <class> <id>
```

Some interpreter commands can also be processed when used as suffixes to a class name:

- `show`: `<class_name> ("<instance_id>")`

Example:

> ```sh
> [FreindlyGadget]: show Brand f0ca205f-31dc-40e4-ac82-09a83d75bcaa
> ```
>
> to print a representation of the Brand with id `f0ca205f-31dc-40e4-ac82-09a83d75bcaa`

> to get the instances of the class `Brand`

- `all`: `<class_name>`

Example:

> ```sh
> [FreindlyGadget]: all Brand
> ```
>
> to print all instances of the class `City`

- `delete`: `<class_name> "<instance_id>"`

Example:

> ```sh
> [FreindlyGadget]: delete Brand "f0ca205f-31dc-40e4-ac82-09a83d75bcaa"
> ```
>
> to delete the instance of class `Brand` with id `f0ca205f-31dc-40e4-ac82-09a83d75bcaa`

- `update`: `<class_name> ("<instance_id>", "<attribute_name>", "<attribute_value>")`

Example:

> ```sh
> [FreindlyGadget]: update Brand ("f0ca205f-31dc-40e4-ac82-09a83d75bcaa", "brand_name", "Infinx")
> ```
>
> to update the attribute called `brand_name` of the instance of `Brand` class with id `f0ca205f-31dc-40e4-ac82-09a83d75bcaa`

The `.update` extension also works with a dictionary of attribute name(s) and value(s)

For a full list of all commands and their usage, run the following command:

```sh
[FreindlyGadget]: help
```

or

```sh
[FreindlyGadget]: help <command>
```

for a help doc on a specific command.

## Usage

1. **Search for a suitable Device:**
   - You will be great with your landing page 
   - Click on “Get Start” and you will be directed to the search section form 
   - Choose your device category (PC or mobile) and start defining your specific purpose and purposes.
   - You can also explore the latest models and view their details in our latest Models

2. **Search and preview:**
     - Matching search results appear and simplified information for the device description is displayed.
     - Use the “Read More” feature to see product details.


## Authors

The [AUTHORS](https://github.com/Yusuf-R/Friendly_Gadget/blob/main/AUTHORS)
file at the root of the repository lists all individuals who were part of the project
from conception. Their full names, links, and contact information are listed below:

<details>
    <summary>Yusuf Abdulwasiu Tunde</summary>
    <ul>
    <li><a href="https://www.github.com/Yusuf-R">GitHub</a></li>
    <li><a href="https://www.linkedin.com/in/abdulwasiu-yusuf-10044299/">Linkedin</a></li>
    <li><a href="mailto:y.abdulwasiu@gmail.com">Gmail</a></li>
    </ul>
</details>
<details>
    <summary>Saria Mohammed</summary>
    <ul>
    <li><a href="https://www.github.com/saria-mohi">GitHub</a></li>
    <li><a href="https://www.linkedin.com/in/saria-mohi-aldein-mohammed-b17630137/">Linkedin</a></li>
    <li><a href="mailto:saria.m.mohammed@gmail.com">Gmail</a></li>
    </ul>
</details>


Fg_Notes/history.txt
import uuid
from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, String
Base = declarative_base()


class Mobile:
    """Base class for gadgets. This class does not map to a table."""
    id = Column(String(60), nullable=False, primary_key=True, default=uuid.uuid4)

    def __init__(self, *args, **kwargs):
      """
      Initialize
      :param args: arguments
      :param kwargs: keyword arguments

      """
      if kwargs:
        for k, v in kwargs.items():
          if k == "__class__":
            continue
          setattr(self, k, v)
        else:
          self.id = str(uuid.uuid4())

    def __str__(self):
        return "[{}] ({}) {}".format(self.__class__.__name__, self.id, self.__dict__)

    def to_dict(self):
        obj_dict = self.__dict__.copy()
        obj_dict["__class__"] = self.__class__.__name__
        if '_sa_instance_state' in obj_dict:
            del obj_dict['_sa_instance_state']
        return obj_dict


from base import Mobile, Base
from sqlalchemy import Column, String
from sqlalchemy.orm import relationship

class Brand(Mobile, Base):
    """Base class for mobiles. This class maps to the 'mobiles' table."""
    __tablename__  = "brands"
    brand_name = Column(String(16), nullable=False)
    models = relationship("Model", backref="brands", cascade="all, delete, delete-orphan")



from base import Mobile, Base
from sqlalchemy import Column, String, ForeignKey
from sqlalchemy.orm import relationship

class Model(Mobile, Base):
    """Subclass of Mobile. This class maps to the 'mobiles' table."""
    __tablename__  = "models"
    model_name = Column(String(32), nullable=False)
    brand_id = Column(String(60), ForeignKey('brands.id'), nullable=False)
    features = relationship("Features", backref="models")



from base import Mobile, Base
from sqlalchemy import Column, String, ForeignKey, JSON
from sqlalchemy.orm import relationship


class Features(Mobile, Base):
  """Subclass of Mobile. This class maps to the 'mobiles' table."""
  __tablename__  = "features"
  model_id = Column(String(60), ForeignKey('models.id'), nullable=False)
  secondary_details = relationship('Secondary', backref='features')
  # primary_details = relationship('Primary', backref='features')

  PhoneDetails = Column(JSON, nullable=True)
  LaunchDetails = Column(JSON, nullable=True)
  NetworkDetails = Column(JSON, nullable=True)
  DisplayDetails = Column(JSON, nullable=True)
  BodyDetails = Column(JSON, nullable=True)
  PlatformDetails = Column(JSON, nullable=True)
  Processor = Column(JSON, nullable=True)
  ChipSet = Column(JSON, nullable=True)
  MemoryDetails = Column(JSON, nullable=True)
  MainCamera = Column(JSON, nullable=True)
  SelfieCamera = Column(JSON, nullable=True)
  Battery = Column(JSON, nullable=True)



from base import Mobile, Base
from sqlalchemy import Column, String, ForeignKey, JSON

class Secondary(Mobile, Base):
    __tablename__ = 'secondary'
    # primary_id = Column(String(60), ForeignKey('primary.id'))
    inner_key = Column(String(256))
    inner_value = Column(String(256))
    feature_id = Column(String(60), ForeignKey('features.id'))



*******************Execution**************************
import urllib.parse
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from brand import Brand
from model import Model, Base
from features import Features
from data import apple_data, samsung_data, spec_iphone_14
from secondary_features import Secondary

usr = "root"
pswd = "Remisql@91"
db = "fg_db"

pd_encd = urllib.parse.quote_plus(pswd)
engine = create_engine('mysql+mysqldb://{}:{}@localhost/{}'.format(usr, pd_encd, db), pool_pre_ping=True, echo=True)
engine.connect()

# Create a session
Session = sessionmaker(bind=engine)
session = Session()

# Create the tables
Base.metadata.create_all(engine)


def create_brands(data):
  for brand in data:
    obj = Brand()
    obj.brand_name = brand
    session.add(obj)
    session.commit()
  session.close()

def create_models(brand_id, data):
  for brand in data:
    for value in brand.values():
        obj = Model()
        obj.model_name = value
        obj.brand_id = brand_id
        session.add(obj)
        session.commit()
  session.close()

def create_features(model_id, data=None):

  feat = [
    "PhoneDetails",
    "LaunchDetails",
    "NetworkDetails",
    "DisplayDetails",
    "BodyDetails",
    "PlatformDetails",
    "Processor",
    "ChipSet",
    "MemoryDetails",
    "MainCamera",
    "SelfieCamera",
    "Battery"
  ]

  if data is None:
    obj = Features()
    obj.model_id = model_id
    session.add(obj)
    session.commit()
    session.close()

  if data is not None:
    obj = Features()
    obj.model_id = model_id
    session.add(obj)
    session.commit()
    for column_data in feat:
      if column_data in data:
        setattr(obj, column_data, data[column_data])
        if isinstance(data[column_data], dict):
          for key, value in data[column_data].items():
            secondary_entry = Secondary()
            secondary_entry.feature_id = obj.id
            secondary_entry.inner_key = key
            secondary_entry.inner_value = value
            session.add(secondary_entry)
    session.commit()
  session.close()


brands = ["Apple", "Huawei", "Samsung", "Tecno", "Nokia"]
create_brands(brands)

apple_id = session.query(Brand).filter_by(brand_name="Apple").first().id
samsung_id = session.query(Brand).filter_by(brand_name="Samsung").first().id


create_models(apple_id, apple_data)
create_models(samsung_id, samsung_data)

# get the id of iphone 14
iphone_14_id = session.query(Model).filter(Model.model_name == "iPhone 14").first().id

create_features(iphone_14_id, spec_iphone_14)






models/main.py
import urllib.parse
from sqlalchemy import create_engine
from sqlalchemy.orm import sessionmaker
from models.brand import Brand
from models.model import Model, Base
from models.features import Features
from models.data import apple_data, samsung_data, spec_iphone_14
from models.secondary_features import Secondary
from models import storage


usr = "root"
pswd = "Remisql@91"
db = "fg_db"
host = "localhost"

pd_encd = urllib.parse.quote_plus(pswd)
engine = create_engine(
    "mysql+mysqldb://{}:{}@{}/{}".format(usr, pd_encd, host, db),
    pool_pre_ping=True,
    echo=False,
)
engine.connect()

# Create a session
Session = sessionmaker(bind=engine)
session = Session()

# Create the tables
Base.metadata.create_all(engine)


clx = {
    "Model": Model,
    "Brand": Brand,
    "Features": Features,
    "Secondary": Secondary,
}


def create_brands(data):
    """Create a new brands"""
    if data == "" or data is None or data == []:
        print("data cannot be empty")
        return
    try:
        if type(data) is str:
            obj = Brand()
            obj.brand_name = data
            session.add(obj)
            session.commit()
            session.close()
            return
        elif type(data) is list:
            for brand in data:
                obj = Brand()
                obj.brand_name = brand
                session.add(obj)
                session.commit()
                session.close()
            return
    except TypeError:
        print("data must be a string or a list of strings")
        return


def create_models(brand_id, data):
    """Create a new model for a particular brand"""
    for brand in data:
        for value in brand.values():
            obj = Model()
            obj.model_name = value
            obj.brand_id = brand_id
            session.add(obj)
            session.commit()
    session.close()


def create_features(model_id, data=None):
    """Populate features for a given model based on it's id"""

    feat = [
        "PhoneDetails",
        "LaunchDetails",
        "NetworkDetails",
        "DisplayDetails",
        "BodyDetails",
        "PlatformDetails",
        "Processor",
        "ChipSet",
        "MemoryDetails",
        "MainCamera",
        "SelfieCamera",
        "Battery" "Purpose",
    ]

    if data is None:
        obj = Features()
        obj.model_id = model_id
        session.add(obj)
        session.commit()
        session.close()

    if data is not None:
        obj = Features()
        obj.model_id = model_id
        session.add(obj)
        session.commit()
        for column_data in feat:
            if column_data in data:
                setattr(obj, column_data, data[column_data])
                if isinstance(data[column_data], dict):
                    for key, value in data[column_data].items():
                        secondary_entry = Secondary()
                        secondary_entry.feature_id = obj.id
                        secondary_entry.inner_key = key
                        secondary_entry.inner_value = value
                        session.add(secondary_entry)
        session.commit()
    session.close()


# brands = ["Apple", "Huawei", "Samsung", "Tecno", "Nokia"]
# create_brands(brands)

# apple_id = session.query(Brand).filter_by(brand_name="Apple").first().id
# samsung_id = session.query(Brand).filter_by(brand_name="Samsung").first().id


# create_models(apple_id, apple_data)
# create_models(samsung_id, samsung_data)

# # get the id of iphone 14

iphone_14_id = (
    session.query(Model).filter(Model.model_name == "iPhone 14").first().id
)

create_features(iphone_14_id, spec_iphone_14)

l = []
brands = storage.all(Model).values()
for data in brands:
    if data.features:
        b = data.features
        for c in b:
            l.append(c.to_dict())
    else:
        continue
print(l)

x = storage.count(Brand)
print(x)


# #create_brands("Infinix")
# # y = storage.all(Brand).values()
# # for x in y:
# #   print(x.to_dict())

# v = storage.get(Brand, "1208902f-b221-42b5-901f-45d02243bdf9")
# print(v.id)
